### PR TITLE
Use CellAssign to label tumor cells in a Ewings sample

### LINKED ISSUE #401: Use CellAssign to classify tumor/normal cells in Ewing sarcoma
### If you are filing this issue based on a specific GitHub Discussion, please link to the relevant Discussion.

#292 

### Describe the goals of the changes to the analysis module.

This issue is to track using `CellAssign` to assign cells as tumor/normal cells in Ewing sarcoma samples. I plan to build a reference of marker genes from the list of tumor marker genes in the `ewings-cell-type` module. This reference will then be used to assign cells from a single sample as tumor/normal. We can then compare the assignments from CellAssign to our manual annotations obtained in #391. 

My original idea was to have two columns, one for tumor cells and one for "other" or normal cells in the marker gene matrix. However, initial exploration in #391 showed that isn't working as expected. We may want to try by adding a few other cell types that are expected to be present (endothelial cells, fibroblasts, etc.) to the reference and see if `CellAssign` prefers that. 

### What will your pull request contain?

This will include a new notebook that is used to prep the reference matrices and evaluate results of running the `run-cellassign.py` script in the module with different references. 

### Will you require additional software beyond what is already in the analysis module?

No, I already set up the script to run `CellAssign`, so this should not require any more software dependencies than are already present. 

### Will you require different computational resources beyond what the analysis module already uses?

No, I should be able to do this locally without any additional resources. 

### If known, when do you expect to file the pull request?

_No response_

### PR DESCRIPTION
**Pull request by allyhawkins**
### Purpose/implementation Section

The purpose of this PR is to explore using `CellAssign` to label tumor and normal cells in one of the Ewing sarcoma samples from SCPCP000015. In particular, I am running `CellAssign` with the tumor marker gene list to see if that list is sufficient for `CellAssign` to classify tumor cells. 

#### Please link to the GitHub issue that this pull request addresses.

Closes #401 

#### What is the goal of this pull request?

The goal is to identify if using `CellAssign` with a reference containing a list of tumor markers is a good method for identifying tumor cells in these samples. I used references that contain just tumor markers and those that combine tumor markers with normal cells. 

#### Briefly describe the general approach you took to achieve this goal.

I ran `CellAssign` on a single sample from SCPCP000015 using four different reference matrices: 

1. Tumor markers only with two options for cell types - tumor or other. 
2. A filtered list of tumor markers that includes only genes with mean gene expression > 1 with two options for cell types - tumor or other. 
3. A combined list of tumor markers and normal markers. The tumor markers are all markers from option 1 with all normal markers coming from the supplemental table in Visser et al.
4. A combined list of tumor markers and some normal markers from PanglaoDB. Markers for endothelial cells and fibroblasts from Panglao were combined with all tumor markers. 

After running `CellAssign` with each reference, I looked at where the cell types being assigned were on the UMAP and compared the assignments to the tumor/normal assignments from #391. 
I also looked at marker gene expression across the groups of cell types and noticed that it did not line up with what I expected. Cells with high expression of tumor marker genes were not classified as tumors and vice versa. 

#### What is your summary of the results?

Overall looking at the results presented in the notebook, it doesn't appear that using `CellAssign` with these smaller reference sets is very good at identifying tumor cells. I think part of it might be the list of tumor markers themselves, we may have too many marker genes, too little, or not the right ones for this sample? It appears that generally `CellAssign` doesn't want to assign cells as tumor cells and when it does these cells do not line up with what I would expect based on the marker gene analysis. 

I think it might be worth trying to add in the marker genes to the bone and soft tissue reference we use for `scpca-nf` and re-running `CellAssign`, but that will take a very long time and a lot of resources so I haven't done that yet. I wanted to see if I could use smaller reference sets and identify tumor cells, but that does not appear to be the case. 
We could also do the same thing but with `mx assign` to see if we can cut down on the computational cost, but that also seemed outside the scope of this PR. 

#### What is the name of your results bucket on S3?

Any results from running `CellAssign` can be found in: 

`s3://researcher-211125375652-us-east-2/ewings-cell-type/results/cellassign_predictions`

#### What types of results does your code produce (e.g., table, figure)?

A TSV file with predictions output from `run-cellassign.py` for each reference used. 

#### What are the software and computational requirements needed to be able to run the code in this PR?

If you download the results from S3 then you should be able to render the notebook without issues. 
To actually run the parts of the notebook that run `CellAssign`, you will need to install and activate the conda environment, `openscpca-ewings-cell-type` and then run the script. 

#### Is there anything that you want to discuss further?

Just a note that this was originally stacked on #391, but I wanted to be able to file the PR to `AlexsLemonade/OpenScPca-analysis` rather than my fork. So for now I'm making this a draft PR and then once #391 goes in, all changes reflected there will be removed from the diff here. At that point I will request review. 

### Author checklists

_Check all those that apply._
_Note that you may find it easier to check off these items after the pull request is actually filed._

#### Analysis module and review

- [x] This analysis module [uses the analysis template and has the expected directory structure](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/analysis-modules/).
- [x] The analysis module `README.md` has been updated to reflect code changes in this pull request.
- [x] The analytical code is documented and contains comments.
- [x] Any results and/or plots this code produces have been added to your S3 bucket for review.

#### Reproducibility checklist

- [ ] Code in this pull request has been added to the GitHub Action workflow that runs this module.
- [ ] The dependencies required to run the code in this pull request have been added to the analysis module `Dockerfile`.
- [x] If applicable, the dependencies required to run the code in this pull request have been added to the analysis module conda `environment.yml` file.
- [x] If applicable, R package dependencies required to run the code in this pull request have been added to the analysis module `renv.lock` file.

### CODE REVIEW

File: analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+readr::write_tsv(filtered_ref_mtx, filtered_markers_mtx_file)
+```
+
+
+```{r, eval=FALSE}
+cellassign_call <- glue::glue(
+  "python scripts/run-cellassign.py",
+  "--anndata_file '{anndata_file}'",
+  "--output_predictions '{filtered_predictions_file}'",
+  "--reference '{filtered_markers_mtx_file}'"
```
Comment (jashapiro) [inline analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd:247]: Even if these spaces are not fully required, I find them helpful...
```suggestion
  " --anndata_file '{anndata_file}'",
  " --output_predictions '{filtered_predictions_file}'",
  " --reference '{filtered_markers_mtx_file}'"
```

File: analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+filtered_celltype_assignments <- get_celltype_assignment(filtered_predictions)
+
+table(filtered_celltype_assignments$celltype)
+```
+
+Now we are seeing a few more tumor cells, let's see where they are on the UMAP. 
+
+```{r}
+# first read in tumor normal classifications 
+classifications_df <- readr::read_tsv(classifications_file)
```
Comment (jashapiro) [inline analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd:268]: This is from your manual classifications? If so, I would probably rename this to clarify.

File: analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+  dplyr::rename_with(
+        \(x) stringr::str_replace(x, "^UMAP\\.", "UMAP"),
+        starts_with("UMAP")
+      ) |> 
+  # join with tumor normal
+  dplyr::left_join(classifications_df) |> 
+  # join with filtered cell type assignments from cellassign 
+  dplyr::left_join(filtered_celltype_assignments, by = c("barcodes" = "barcode"))
+
+ggplot(celltype_df, aes(x = UMAP1, UMAP2, color = filtered_cellassign_annotations)) +
```
Comment (jashapiro) [inline analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd:285]:
```suggestion
ggplot(celltype_df, aes(x = UMAP1, y = UMAP2, color = filtered_cellassign_annotations)) +
```

**Comment by allyhawkins** [comment]
Just noting that I did just double check these and the reference matrices are correct with each cell type having the correct genes listed. I also went through the `run-cellassign.py` line by line with one of the H5 files and the Visser reference and the reference we provide is correct and read in correctly... Also we subset the AnnData object to contain only genes in the reference matrix and that looks correct. 

I also double checked that all the Ensembl IDs were correct.

File: analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+readr::write_tsv(panglao_tumor_ref_mtx, panglao_markers_mtx_file)
+```
+
+
+```{r, eval=FALSE}
+cellassign_call <- glue::glue(
+  "python scripts/run-cellassign.py ",
+  "--anndata_file '{anndata_file}' ",
+  "--output_predictions '{panglao_predictions_file}' ",
+  "--reference '{panglao_markers_mtx_file}' "
```
Comment (jashapiro) [inline analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd:584]: This one you had spaces at the end, so I would expect it to work, but I moved them to the beginning for consistency
```suggestion
  "python scripts/run-cellassign.py",
  " --anndata_file '{anndata_file}'",
  " --output_predictions '{panglao_predictions_file}'",
  " --reference '{panglao_markers_mtx_file}'"
```

File: analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd [hunk 5]
```diff
@@ -0,0 +1,641 @@
...
+readr::write_tsv(visser_ref_mtx, visser_markers_mtx_file)
+```
+
+
+```{r, eval=FALSE}
+cellassign_call <- glue::glue(
+  "python scripts/run-cellassign.py ",
+  "--anndata_file '{anndata_file}' ",
+  "--output_predictions '{visser_predictions_file}' ",
+  "--reference '{visser_markers_mtx_file}'"
```
Comment (jashapiro) [inline analyses/ewings-cell-type/02-cellassign-tumor-cell-classification.Rmd:379]: Move spaces to front for conisistency
```suggestion
  "python scripts/run-cellassign.py",
  " --anndata_file '{anndata_file}'",
  " --output_predictions '{visser_predictions_file}'",
  " --reference '{visser_markers_mtx_file}'"
```

**Review by jashapiro — APPROVED** [review]
This looks good to me. I spent some time looking through the code and the underlying results trying to find where you might have an error that might result in misclassification, and I couldn't find anything! 

The only thing I might suggest is a heatmap of the raw cellassign scores, labelled also with the manual classifications. But having looked at the raw results I don't expect much out of that that woudl change anything.

**Comment by allyhawkins** [comment]
> The only thing I might suggest is a heatmap of the raw cellassign scores, labelled also with the manual classifications. But having looked at the raw results I don't expect much out of that that woudl change anything.

I ended up adding these in because I think they are helpful to see for the classifications where we are using both tumor and normal. But they don't really change the results or overall findings. Tumor cells are still not getting called as tumor cells.

**Pull request merged**